In [1]:
# import libraries
import pandas as pd
import numpy as np
import pickle
import random

In [2]:
with open('full_process.pkl', 'rb') as file:
  full_process = pickle.load(file)

In [3]:
#membuat data dummy untuk data inferences
np.random.seed(3)
n = 100
oldpeak = np.random.uniform(0,6.2,n)
chol = np.random.uniform(126,564,n)
thall =[0, 1, 2, 3]
thall = random.choices(thall, k=n)
cp = [0, 1, 2, 3]
cp = random.choices(cp, k=n)
exng = [0, 1]
exng = random.choices(exng, k=n)
sex= [0, 1]
sex = random.choices(sex, k=n)
caa= [0, 1, 2, 3,4]
caa = random.choices(caa, k=n)
slp= [0, 1, 2]
slp = random.choices(slp, k=n)


In [4]:
data_inf = pd.DataFrame({
   'oldpeak' : oldpeak,
   'chol' : chol,
   'thall' : thall ,
   'cp' : cp,
   'exng' : exng ,
   'sex' : sex,
   'caa' : caa,
   'slp' : slp
})

data_inf

oldpeak        chol  thall  cp  exng  sex  caa  slp
0   3.414947  210.127666      0   3     1    1    1    2
1   4.390517  443.843295      3   2     0    0    4    1
2   1.803609  469.802276      1   0     0    1    4    2
3   3.167131  551.779083      2   0     0    1    2    2
4   5.536271  498.725481      1   0     0    1    2    0
..       ...         ...    ...  ..   ...  ...  ...  ...
95  5.666647  158.428555      3   2     1    0    0    0
96  3.928726  335.764968      2   0     1    1    0    2
97  2.268832  153.278412      3   2     0    1    3    0
98  3.427636  513.297747      3   3     0    0    1    0
99  1.217560  321.264858      0   1     0    1    4    2

[100 rows x 8 columns]

In [5]:
y_pred_inf = full_process.predict(data_inf)
y_pred_inf

c:\Users\USER\miniconda3\lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


array([1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0], dtype=int64)

In [6]:
# Mengkonversi ke dataframe

y_pred_inf_df = pd.DataFrame(y_pred_inf, columns=['Heart Attack - Prediction'])
y_pred_inf_df

Heart Attack - Prediction
0                           1
1                           0
2                           0
3                           0
4                           0
..                        ...
95                          0
96                          0
97                          0
98                          0
99                          0

[100 rows x 1 columns]

In [7]:
#mengconcat data inference dan hasil predict
tabel_inference = pd.concat([data_inf, y_pred_inf_df], axis=1)
tabel_inference

oldpeak        chol  thall  cp  exng  sex  caa  slp  \
0   3.414947  210.127666      0   3     1    1    1    2   
1   4.390517  443.843295      3   2     0    0    4    1   
2   1.803609  469.802276      1   0     0    1    4    2   
3   3.167131  551.779083      2   0     0    1    2    2   
4   5.536271  498.725481      1   0     0    1    2    0   
..       ...         ...    ...  ..   ...  ...  ...  ...   
95  5.666647  158.428555      3   2     1    0    0    0   
96  3.928726  335.764968      2   0     1    1    0    2   
97  2.268832  153.278412      3   2     0    1    3    0   
98  3.427636  513.297747      3   3     0    0    1    0   
99  1.217560  321.264858      0   1     0    1    4    2   

    Heart Attack - Prediction  
0                           1  
1                           0  
2                           0  
3                           0  
4                           0  
..                        ...  
95                          0  
96                          0  
97                          0  
98                          0  
99                          0  

[100 rows x 9 columns]

# Kesimpulan : 

## Conceptual Problems

1. Jelaskan latar belakang adanya bagging dan cara kerja bagging !

Bagging dibuat untuk mengatasi masalah overfitting dan meningkatkan kinerja model. Baging muncul karena saat kita mengerjakan machine learning kita memiliki dataset yang kurang bagus sehingga menghasilkan model yang kurang baik sehingga bagging dibuat untuk mengatasi masalah tersebut.

Cara kerjanya : Bagging pertama-tama membuat beberapa set data latihan baru (bootstrap samples) dengan mengambil sampel acak dengan penggantian dari dataset asli . Lalu model difit ke booststrap sample , lalu hasil prediksi dari semua model dihasilkan dan kemudian diagregasi. Hasil prediksi digabungkan dengan cara mengambil rata-rata(untuk regresi) dan voting (untuk klasifikasi)

2. Jelaskan perbedaan cara kerja algoritma Random Forest dengan algoritma boosting yang Anda pilih !

Random Forest : menggunakan bagging untuk membangun beberapa pohon keputusan independen dengan pemilihan fitur acak, dan hasil prediksinya diambil berdasarkan rata-rata (regresi) atau mode (klasifikasi) dari pohon-pohon tersebut untuk mengurangi overfitting.

Boosting : teknik ensemble di mana setiap model berikutnya berusaha untuk mengoreksi kesalahan model sebelumnya dengan memberikan bobot lebih kepada data yang salah diprediksi, sehingga secara bertahap meningkatkan performa keseluruhan model.

3. Jelaskan apa yang dimaksud dengan Cross Validation !

Cross-validation adalah sebuah teknik yang digunakan dalam pembangunan model machine learning untuk mengevaluasi kinerja model dengan cara membagi dataset menjadi beberapa subset atau fold. Tujuannya adalah untuk menghindari overfitting dan memberikan tahu model mana yang akan memberikan hasil prediksi yang lebih baik

#### Problem stament bisa terpenuhi karena recall score untuk meminimalisir false negative bisa dilakukan sebesar 97%

### Kelebihan dan Kekurangan Model SVM Random Search

Karena model svm memiliki nilai recal score range test set yang paling besar maka saya memilih model SVM yaitu

** Kelebihan :

- Nilai score yang tinggi masih berada dalam perkiraan range test set score

Recall Score - Range of Test-Set       :  0.8067028346993861 - 0.9784823504857993

- Memiliki nilai STD yang rendah 0.07 mendekati 0  dan mean yang cukup tinggi 0.86 mendekati 1

- Nilai train dan test model SVM Randomsearch yang tidak jauh berbeda dan cenderung dekat

Recall Score - Train Set  :  0.9541984732824428 

Recall Score - Test Set  :  0.9696969696969697 

Recall score yang tinggi 0.96 berarti bisa meminimalisir 96% data false negative (recall)

** Kekurangan :

- Masih ada potensi error walaupun angkanya kecil hanya 1 / 3% di confusion matrix SVM Random Search


### Saran perbaikan untuk model SVM RandomSearch : 

- Memperbaiki pemilihan feature selection agar memastikan fitur - fitur yang dipilih adalah fitur yang terbaik / paling relevan 

- Pemilihan Parameter bisa ditingkatkan lagi dengan menambah parameter atau memilih parameter yang lebih baik untuk meningkatkan nilai

- Penggunaan hyperparameter tuning mungkin bisa menggunakan manual search atau grid search untuk memiliki nilai yang lebih baik

### Kesimpulan EDA 

- Pasien yang terkena serangan jantung lebih kecil daripada yang tidak terkena serangan jantung. Kita bisa memberikan imbauan sederhana pada setiap pasien yang berobat untuk menjaga pola kesehatan pada pasien yang memiliki agar tidak terkena serangan jantung

- Rata-Rata Umur pasien lebih dari 40 tahun sampai 65 tahun dan pasien terbanyak ada diumur 55-60 tahun sehingga pasien pada umur tersebut. Umur yang semakin tua akan lebih rentan terkena serangan jantung walau tidak menutup kemungkinan yang lebih muda bisa terkena serangan jantung juga

- Pasien yang memiliki angina tipikal(nyeri dada yang berat) maka akan memiliki nyeri dada ketika berolahraga. Semakin rendah tingkat nyeri dada maka munculnya nyeri dada ketika berolahraga semakin sedikit. 

    Saran : Untuk pasien yang memiliki angina tipikal disarankan tidak melakukan aktifitas olahraga yang berat mungkin bisa disarankan dengan olahraga jalan kaki saja untuk mengurangi ,unculnya nyeri dada saat berolahraga

- Semakin besar / semakin banyak penyakit dalam pembuluh darah maka presentasi pasien terkena gula darah semakin besar . Jika dilihat dari yang index 0 jumlah penyakit pembuluh darah yang sedikit maka prentasi pasien yang terkena gula darah semakin rendah begitu juga sebaliknya.

    Saran : Orang yang memiliki penyakit di pembuluh darah yang banyak lebih baik disarankan untuk diet gula karena kemungkinan besar pasien terkena diabetes yang bisa membuat orang terkena penyakit jantung

- Pria memiliki detak jantuk yang tinggi dibandingkan wanita , walaupun ada wanita yang memiliki detak jantung sampai 200 itu itu dikarenakan pria lebih suka beraktivitas berat dibandingkan dengan wanita sehingga detak jantung pria berdetak lebih cepat dibandingkan dengan wanita

    Saran : Untuk pria bisa dipertahankan dengan aktivitas fisik yang sesuai dengan kesahatan fisik dan untuk wanita bisa ditingkatkan untuk aktivitas fisiknya sesuai dengan kondisi fisik untuk meningkatkan kesahatan jantung

- Wanita lebih banyak terkena serangan jantung dibandingkan pria , akan tetapi jika dilihat secara keseluruhan pria memiliki potensi terkena serangan jantung dibandingkan dengan wanita . Karena jumlah pria yang tidak terkena jantung berbanding cukup jauh lebih sedikit dibandingkan dengan yang terkena serangan jantung . Sehingga presentase pria yang terkena serangan jantung lebih besar daripada 
yang tidak terkena serangan jantung

    Saran : Untuk Pria lebih menjaga pola makan dan pola hidup karena kemungkinan terkena serangan jantung lebih besar , tetapi perempuan harus menjaga pola hidup dan menjaga pola makan juga karena jumlah yang terkena serangan jantung lebih besar dari pria

- Dilihat bahwa orang yang memiliki kolestrol yang semakin tinggi maka tekanan darahnya akan semakin tinggi sehingga disarankan bagi orang yang memiliki kolestrol yang tinggi sudah harus menjaga pola makan untuk menurunkan tekanan darah dalam istirahat

    Saran : pasien dengan kadar kolestrol dan tekanan darah yang tinggi bisa mengurangi konsumsi garam , kafein , dan alkohol. Pasien juga bisa meminum obat kolstrol dan obat darah tinggi bila memiliki kadar kolestrol dan tekanan darah yang tinggi

Saran pola hidup jika pasien terkena serangan jantung :

- Berolahraga teratur

- Mengonsumsi makanan yang menyehatkan jantung

- Menurunkan tekanan darah

- Menjaga berat badan yang sehat

- Berfokus pada kesehatan mental

- Kelola stres

- Berhenti merokok

Sumber :
https://lifestyle.kompas.com/read/2021/11/09/101744620/7-pola-hidup-sehat-setelah-mengalami-serangan-jantung?page=all

Saran pola hidup jika pasien tidak terkena serangan jantung dan untuk mencegah serangan jantung

- Konsumsi Makanan Ramah Jantung

- Meditasi 

- Berhenti Merokok

- Aktif Bergerak dan Berolahraga

- Cek Kesehatan Secara Rutin

Sumber : https://www.siloamhospitals.com/informasi-siloam/artikel/cara-cegah-serangan-jantung-di-usia-muda